Add BentoManga connector

TristanWasTaken opened the pull request:
I ~~stole~~ borrowed the [connector](https://github.com/manga-download/hakuneko/issues/4513#issuecomment-1086654730) from @courli79 and fixed the out of date `_getChapters` function.
This solves #5114 and kinda solves #4513

The PR is marked as draft since, as it turns out, BentoManga now has an api
Example: https://bentomanga.com/manga/poison-dragon-the-legend-of-an-asura/chapter/56
- https://bentomanga.com/api/?id=237487&type=chapter
- https://bentomanga.com/api/?id=10202&type=manga

The connector url needs improvement too since BentoManga has a `www` subdomain too
Example:
- https://bentomanga.com/manga/poison-dragon-the-legend-of-an-asura
- https://www.bentomanga.com/manga/poison-dragon-the-legend-of-an-asura

There is also the chapter page problem. BentoManga gets more chapters by requesting the list html in a similar way as the chapter api requests.
example: https://bentomanga.com/manga/everyone-else-is-a-returnee
- https://bentomanga.com/manga/everyone-else-is-a-returnee?order=desc&limit=1
- https://bentomanga.com/manga/everyone-else-is-a-returnee?order=desc&limit=2

**Tasks:**
- [x] implement api
- [x] add url regex
- [x] solve _getChapter page limit

Fixes https://github.com/manga-download/hakuneko/issues/5114
Fixes https://github.com/manga-download/hakuneko/issues/4513 (Japanread is now bentomanga, its redirecting)

TristanWasTaken committed:
BentoManga: add connector

Much thanks to @courli79 for the connector (https://github.com/manga-download/hakuneko/issues/4513#issuecomment-1086654730)
Pretty much everything except the _getChapters function still works.

Co-Authored-By: courli79 <82915245+courli79@users.noreply.github.com>

TristanWasTaken committed:
BentoManga: fix _getChapters querySelector

MikeZeDev commented on src/web/mjs/connectors/BentoManga.mjs:
```diff
@@ -0,0 +1,79 @@
+
+    async _getMangaFromURI(uri) {
+        let request = new Request(uri, this.requestOptions);
+        let data = await this.fetchDOM(request, 'h1'); 
```
Way too lax CSS selector. You must make sure it will be unique. 

``` div.component-manga-title_main h1 ``` is a better candidate

MikeZeDev commented on src/web/mjs/connectors/BentoManga.mjs:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    async _getMangas() {
+        let request = new Request(this.url + '/manga_list', this.requestOptions); 
```
This method is fine by itself but Bentomanga got more than 1k manga pages. 
Suggestion to use _getmangas and _getMangasFromPage from [Sample.mjs](https://github.com/manga-download/hakuneko/blob/master/src/web/mjs/connectors/templates/%40Samples.mjs)

If you want to be efficient, you can get JSON manga with html data inside by calling 
https://bentomanga.com/manga_list?limit=X&cb=Y . 

X being page number-1, cb being the timestamp from when you call the page.
This request must get a special header, with name "X-Requested-With" and value "XMLHttpRequest". 

Then you get some html in the json ("mangas")

```
	<div class="div-manga_cover">
		<a href="https://bentomanga.com/manga/maseki-gurume-mamono-no-chikara-o-tabeta-ore-wa-saikyou" class="component-manga-cover">
			<img alt="couverture manga" src="https://bentomanga.com/media/cache/manga_medium/images/mangas/cover/5e47469782828797082475.jpg">
				<span class="div-manga_cover-title">Maseki Gurume: Mamono no Chikara o Tabeta Ore wa Saikyou!</span>
		</a>
		<span class="badge-chapterCount mobile-adapt">0/29</span>
	</div>
```

There are plenty of sample about how to do that in other connectors.


        request.headers.set('X-Requested-With', 'XMLHttpRequest'); 
        const data = await this.fetchJSON(request);
        const dom =  this.createDOM( data.mangas ); //json got html  data in "mangas"
        const mangaNode = dom.querySelector('div.div-manga_cover a');
        return mangaNode.map(element => {
            return {
                id: this.getRootRelativeOrAbsoluteLink(element, this.url),
                title: element.querySelector('span.div-manga_cover-title').textContent.trim()
            };
        });

MikeZeDev commented on src/web/mjs/connectors/BentoManga.mjs:
```diff
@@ -0,0 +1,79 @@
+        return data.map(element => {
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element, request.url),
+                title: element.querySelector('span.chapter_volume').innerText.trim() + ' ' + element.querySelector('span.chapter_title').innerText.trim(),
```
Are there other elements than chapter_volume and chapter_title ? cant we just get element.text.trim() as title? 
Not a big deal tbh

MikeZeDev commented on src/web/mjs/connectors/BentoManga.mjs:
```diff
@@ -0,0 +1,79 @@
+                const info = document.querySelector('head meta[data-chapter-id]');
+                const uri = new URL('/api/?type=chapter&id=' + info.dataset.chapterId, window.location.origin);
+                const customHeaders = {
+                    headers: {"a":"Math.random().toString(16)"}
```
Apparently, no need for "a" in custom headers. But you will need ('X-Requested-With', 'XMLHttpRequest') like before; 

``` headers : { 'X-Requested-With' : 'XMLHttpRequest' } ```

MikeZeDev reviewed (changes requested):
* Manga list should be multipages like on the website
* Manga list could be fetched using api to reduce server load
* Must fix clipboard selector
* You said yourself there are manga with www and without www in url so you must override `canHandleUri` and use a good regex that match both cases

TristanWasTaken committed:
BentoManga: make _getMangaFromURI css selector unique

Co-Authored-By: MikeZeDev <10671518+MikeZeDev@users.noreply.github.com>

TristanWasTaken committed:
BentoManga: add 'X-Requested-With' header

TristanWasTaken committed:
BentoManga: add _getListPage function

TristanWasTaken committed:
BentoManga: cleanup _getListPage

MikeZeDev commented:
I think you are wasting your time because you said Cloudflare blocked you. No point adding a connector that theorically works id no one can use it ;-;

TristanWasTaken committed:
BentoManga: list all chapters

TristanWasTaken committed:
BentoManga: list all manga

TristanWasTaken committed:
BentoManga: cleanup code

TristanWasTaken committed:
BentoManga: fix empty chapter list if pageCount is NaN

TristanWasTaken committed:
BentoManga: fix manga url

Directly calling a list page with the query param "?limit=x" is possible too. This would confuse the _getChapters() logic.
Fix: remove the query params from the manga url.

TristanWasTaken committed:
BentoManga: remove _getPages() captcha check

TristanWasTaken committed:
BentoManga: fix lint errors

- L69-L74: Replace if logic with simple check
- L82: return null instead of undefined.

TristanWasTaken committed:
BentoManga: lint cleanup

TristanWasTaken committed:
BentoManga: remove unessecary null

TristanWasTaken committed:
BentoManga: add url regex

TristanWasTaken committed:
BentoManga: cleanup manga id

TristanWasTaken committed:
BentoManga: cleanup chapter id

TristanWasTaken committed:
BentoManga: fix manga id

MikeZeDev committed:
Update BentoManga.mjs

ronny1982 commented:
[Missing in HaruNeko](https://github.com/search?q=repo%3Amanga-download%2Fharuneko%20%27bentomanga%27&type=code)
